### PR TITLE
[LLVM][NVPTX] Add support for div.full instruction

### PR DESCRIPTION
**Pull request by schwarzschild-radius**
This commit adds NVPTX support for div.full PTX instruction with test under div.ll. [For more information, see PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/#floating-point-instructions-div)

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td [hunk 1]
```diff
@@ -1096,6 +1096,18 @@ def INT_NVVM_DIV_RM_D : F_MATH_2<"div.rm.f64 \t$dst, $src0, $src1;",
 def INT_NVVM_DIV_RP_D : F_MATH_2<"div.rp.f64 \t$dst, $src0, $src1;",
   Float64Regs, Float64Regs, Float64Regs, int_nvvm_div_rp_d>;
 
+def : Pat<(int_nvvm_div_full Float32Regs:$a, Float32Regs:$b),
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1099]: Any particular reason not to use `F_MATH_2` as the other `nvvm_div_*` intrinsics do?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1099]: One particular problem this patch may run into is that `FDIV32ri*` records are predicated on `do_DIVF32_FULL`, and that would be affected by the command line option `nvptx-prec-divf32`. For the intrinsic, you want to emit `div.full` unconditionally.

Comment (schwarzschild-radius) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1099]: @Artem-B I tried setting `--nvptx-prec-divf32=`  to different values and it does not fail. It only failed after adding the same `Requires` clause to the pattern that I notice the Instruction Selection to fail. Seems like Patterns don't inherit the `Requires` clause. What do you think?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1099]: You're right. Per-instruction predicate is for it's own pattern. You're OK here.

File: llvm/test/CodeGen/NVPTX/div.ll [hunk 2]
```diff
@@ -0,0 +1,10 @@
+; RUN: llc < %s -march=nvptx64 | FileCheck %s
+; RUN: %if ptxas %{ llc < %s -march=nvptx64 | %ptxas-verify %}
+
```
[the remark this answers is not in the document]
Comment (schwarzschild-radius) [reply llvm/test/CodeGen/NVPTX/div.ll:3]: Added tests for immediate variants

File: llvm/test/CodeGen/NVPTX/div.ll [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+define float @div_full(float %a, float %b) {
+  ; CHECK: div.full.f32 {{%f[0-9]+}}, {{%f[0-9]+}}, {{%f[0-9]+}}
+  %1 = call float @llvm.nvvm.div.full(float %a, float %b)
+  ; CHECK: div.full.ftz.f32 {{%f[0-9]+}}, {{%f[0-9]+}}, {{%f[0-9]+}}
+  %2 = call float @llvm.nvvm.div.full.ftz(float %1, float %b)
+  ret float %2
+}
```
[the remark this answers is not in the document]
Comment (schwarzschild-radius) [reply llvm/test/CodeGen/NVPTX/div.ll]: Added new line to the file

File: llvm/test/CodeGen/NVPTX/div.ll [hunk 4]
```diff
@@ -0,0 +1,16 @@
+; RUN: llc < %s -march=nvptx64 | FileCheck %s
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/div.ll:1]: Generating checks with utils/update_llc_test_checks.py would be a good idea.

**Review by Artem-B — APPROVED** [review]
LGTM w/ a test nit.

File: llvm/test/CodeGen/NVPTX/div.ll (hunk 4 above)
Comment (schwarzschild-radius) [reply llvm/test/CodeGen/NVPTX/div.ll:2]: Thanks @Artem-B for the suggestion! I have updated the test by running `utils/update_llc_test_checks.py`. Can I go ahead with the merge?

**Review by Artem-B — APPROVED** [review]

File: llvm/test/CodeGen/NVPTX/div.ll (hunk 4 above)
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/div.ll:2]: Go ahead.

**Pull request merged**